### PR TITLE
FaceSkeleton Vroom 1

### PR DESCRIPTION
**Pull request by AbdAlazezAhmed**
Makes face skeleton construction faster by using topology information.
Benchmarks:
```julia
julia> using Ferrite

julia> using BenchmarkTools

julia> grid = generate_grid(Hexahedron,(100,100,100))
Grid{3, Hexahedron, Float64} with 1000000 Hexahedron cells and 1030301 nodes

julia> topology = ExclusiveTopology(grid);

julia> @benchmark Ferrite. _faceskeleton(topology, grid)
BenchmarkTools.Trial: 19 samples with 1 evaluation.
 Range (min … max):  221.792 ms … 410.902 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     254.305 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   271.663 ms ±  46.478 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▁▁   ██▁█ ▁  █      █▁    ▁   ▁        ▁                    ▁  
  ██▁▁▁████▁█▁▁█▁▁▁▁▁▁██▁▁▁▁█▁▁▁█▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  222 ms           Histogram: frequency by time          411 ms <

 Memory estimate: 46.23 MiB, allocs estimate: 2.
 ```
 
 Current implementation:
 ```julia
 julia> using Ferrite

julia> using BenchmarkTools

julia> grid = generate_grid(Hexahedron,(100,100,100))
Grid{3, Hexahedron, Float64} with 1000000 Hexahedron cells and 1030301 nodes

julia> topology = ExclusiveTopology(grid);

julia> @benchmark Ferrite. _faceskeleton(topology, grid)
BenchmarkTools.Trial: 3 samples with 1 evaluation.
 Range (min … max):  1.578 s …    1.932 s  ┊ GC (min … max): 0.00% … 18.59%
 Time  (median):     1.763 s               ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.758 s ± 176.799 ms  ┊ GC (mean ± σ):  6.81% ± 10.74%

  █                             █                          █  
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  1.58 s         Histogram: frequency by time         1.93 s <

 Memory estimate: 652.91 MiB, allocs estimate: 17026973.
 ```
 `grid` argument is kept in case it's needed in the future. The `if-else` thing can be avoided if we consistently define "faces" as vertices for 1D case.

### CODE REVIEW

File: src/Grid/topology.jl [hunk 1]
```diff
@@ -332,21 +332,24 @@ Creates an iterateable face skeleton. The skeleton consists of `FaceIndex` that
...
-    for (cellid,cell) ∈ enumerate(grid.cells)
-        for (local_face_id,face) ∈ enumerate(faces(cell))
-            push!(face_skeleton_global, sortface_fast(face))
-            fs_length_new = length(face_skeleton_global)
-            if fs_length != fs_length_new
-                push!(face_skeleton_local, FaceIndex(cellid,local_face_id))
-                fs_length = fs_length_new
-            end
+    i = 1;
+    if isempty(top.face_face_neighbor)
```
[the remark this answers is not in the document]
Comment (AbdAlazezAhmed) [reply src/Grid/topology.jl]: New banchmarks:
```julia
julia> @benchmark Ferrite. _faceskeleton(topology, grid)
BenchmarkTools.Trial: 26 samples with 1 evaluation.
 Range (min … max):  182.499 ms … 208.627 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     193.976 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   192.880 ms ±   5.542 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

                             █                                   
  ▅▁▁▁▁▁▅█▅▁▁▅▁▁▅▁▅▁▁▁▁▁▅▁█▅▅█▁▅█▁▁▅▅▁▅▁▁▁▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅ ▁
  182 ms           Histogram: frequency by time          209 ms <

 Memory estimate: 46.23 MiB, allocs estimate: 2.
 ```

**Review by termi-official — APPROVED** [review]
LGTM

**Comment by AbdAlazezAhmed** [comment]
Trying to improve performance a bit more
```julia
julia> @benchmark Ferrite. _faceskeleton(topology, grid)
BenchmarkTools.Trial: 39 samples with 1 evaluation.
 Range (min … max):  116.631 ms … 164.498 ms  ┊ GC (min … max): 0.00% … 28.98%
 Time  (median):     129.052 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   128.925 ms ±   6.938 ms  ┊ GC (mean ± σ):  0.95% ±  4.64%

                 █▁                                              
  ▄▁▃▁▁▁▁▁▁▁▁▁▁▁▁██▆▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃ ▁
  117 ms           Histogram: frequency by time          164 ms <

 Memory estimate: 46.23 MiB, allocs estimate: 2.
 ```
 Maybe we can also get rid of the extra allocation (and GC?)

**Comment by termi-official** [comment]
Can you resolve the conflict? Should be good to go then. 10x improvement is really nice. :)

File: src/Grid/topology.jl [hunk 2]
```diff
@@ -330,19 +330,16 @@ Creates an iterateable face skeleton. The skeleton consists of `FaceIndex` that
...
+    cells = getcells(grid)
+    cell_dim = getdim(first(cells))
+    @assert all(cell -> getdim(cell) == cell_dim, cells) "Face skeleton construction requires all the elements to be of the same dimensionality"
+    i = 1
+    neighborhood = cell_dim == 1 ? top.vertex_vertex_neighbor : top.face_face_neighbor
+    face_skeleton_local = Array{FaceIndex}(undef, length(neighborhood) - count(neighbors -> !isempty(neighbors) , neighborhood) ÷ 2)
+    for (idx, face) in pairs(neighborhood)
+        isempty(face.neighbor_info) || face.neighbor_info[][1] > idx[1] || continue
+        face_skeleton_local[i] = FaceIndex(idx[1], idx[2])
+        i+=1
```
Comment (koehlerson) [inline src/Grid/topology.jl:342]: just cosmetics
```suggestion
        i += 1
```

**Review by koehlerson — APPROVED** [review]

**Pull request merged**